ext/process_tags.c

```diff
@@ -0,0 +1,356 @@
+    if (last_backslash && (!last_slash || last_backslash > last_slash)) {
+        last_slash = last_backslash;
+    }
+#endif
+
+    if (last_slash) {
+        *last_slash = '\0';
+        
+        // Get the basename of the directory
+        char *basedir = get_basename(path_copy);
+        free(path_copy);
+        return basedir;
+    }
+
+    free(path_copy);
+    return strdup(".");
+}
+
+/**
+ * Add a process tag entry
+ * 
+ * @param key The tag key (will be duplicated)
+ * @param value The tag value (will be normalized and duplicated)
+ */
+static void add_process_tag(const char *key, const char *value) {
+    if (!key || !value || process_tags.count >= MAX_PROCESS_TAGS) {
+        return;
+    }
+
+    char *normalized_value = normalize_value(value);
+    if (!normalized_value) {
+        return;
+    }
+
+    process_tags.entries[process_tags.count].key = strdup(key);
+    process_tags.entries[process_tags.count].value = normalized_value;
+    
+    if (process_tags.entries[process_tags.count].key) {
+        process_tags.count++;
+    } else {
+        free(normalized_value);
+    }
+}
+
+/**
+ * Comparison function for qsort to sort process tags by key
+ */
+static int compare_tags(const void *a, const void *b) {
+    const process_tag_entry_t *tag_a = (const process_tag_entry_t *)a;
+    const process_tag_entry_t *tag_b = (const process_tag_entry_t *)b;
+    return strcmp(tag_a->key, tag_b->key);
+}
+
+/**
+ * Serialize process tags into a comma-separated string
+ * Format: key1:value1,key2:value2,...
+ * Keys are sorted alphabetically
+ */
+static void serialize_process_tags(void) {
+    if (process_tags.count == 0) {
+        return;
+    }
+
+    // Sort tags by key
+    qsort(process_tags.entries, process_tags.count, sizeof(process_tag_entry_t), compare_tags);
+
+    // Calculate total length needed
+    size_t total_len = 0;
+    for (size_t i = 0; i < process_tags.count; i++) {
+        total_len += strlen(process_tags.entries[i].key);
+        total_len += 1; // for ':'
+        total_len += strlen(process_tags.entries[i].value);
+        if (i < process_tags.count - 1) {
+            total_len += 1; // for ','
+        }
+    }
+
+    // Allocate and build the serialized string
+    char *buffer = (char *)malloc(total_len + 1);
+    if (!buffer) {
+        return;
+    }
+
+    char *ptr = buffer;
+    for (size_t i = 0; i < process_tags.count; i++) {
+        size_t key_len = strlen(process_tags.entries[i].key);
+        size_t value_len = strlen(process_tags.entries[i].value);
+        
+        memcpy(ptr, process_tags.entries[i].key, key_len);
+        ptr += key_len;
+        *ptr++ = ':';
+        memcpy(ptr, process_tags.entries[i].value, value_len);
+        ptr += value_len;
+        
+        if (i < process_tags.count - 1) {
+            *ptr++ = ',';
+        }
+    }
+    *ptr = '\0';
+
+    // Create a persistent zend_string
+    process_tags.serialized = zend_string_init(buffer, total_len, 1);
```

Comment:
The better way would be

```
    process_tags.serialized = zend_string_alloc(total_len, 1);
    char *ptr = ZSTR_VAL(process_tags.serialized);
```
rather than creating a temporary buffer, to copy and free it.